### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by backslashbaker**

### CODE REVIEW

File: spec/game_spec.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
+require './lib/game'
+require './lib/display'
+require './lib/board'
+
+
```
Comment (MollieS) [inline spec/game_spec.rb]: 👾 - again you can leave 2 newlines between require statements and class definitions, but be consistent

File: spec/game_spec.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    allow(display_spy).to receive(:repeat?).and_return(false)
+    game = Game.new(display_spy, board_spy)
+
+    game.run
+
+    expect(display_spy).to have_received(:display_board).with(grid).twice
+  end
+
+  it 'gets user input for x coordinate' do
+    allow(display_spy).to receive(:repeat?).and_return(false)
```
Comment (MollieS) [inline spec/game_spec.rb]: You could use a before or before(:each) block to remove this duplication

File: spec/game_spec.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    allow(display_spy).to receive(:get_x).and_return(1)
+    allow(display_spy).to receive(:get_y).and_return(1)
+    allow(display_spy).to receive(:repeat?).and_return(false)
+
+    game = Game.new(display_spy, board_spy)
+
+    game.run
+
+    expect(board_spy).to have_received(:mark_board)
+
```
Comment (MollieS) [inline spec/game_spec.rb]: 👾

File: spec/game_spec.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    allow(board_spy).to receive(:grid).and_return(grid)
+    allow(display_spy).to receive(:get_x).and_return(1)
+    allow(display_spy).to receive(:get_y).and_return(1)
+    allow(display_spy).to receive(:repeat?).and_return(false)
+
+    game = Game.new(display_spy, board_spy)
+
+    game.run
+
+    expect(board_spy).to have_received(:mark_board)
```
Comment (MollieS) [inline spec/game_spec.rb]: You could make this expectation more meaningful by asserting on the arguments passed to the mark_board method - you've already set up that they should be 1 and 1 on line 52 & 53

The message expectations section here might be useful: https://relishapp.com/rspec/rspec-mocks/docs

File: spec/game_spec.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    allow(board_spy).to receive(:grid).and_return(grid)
+    allow(display_spy).to receive(:get_x).and_return(1, 2)
+    allow(display_spy).to receive(:get_y).and_return(1, 2)
+    allow(display_spy).to receive(:repeat?).and_return(true, false)
+
+    game = Game.new(display_spy, board_spy)
+
+    game.run
+
+    expect(board_spy).to have_received(:mark_board).at_least(2).times
```
Comment (MollieS) [inline spec/game_spec.rb]: You could also add the expected arguments here

File: spec/game_spec.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    game = Game.new(display_spy, board_spy)
+
+    game.run
+
+    expect(display_spy).to have_recieved(:error)
+    expect(board_spy).to_not have_received(:mark_board)
+  end
+end
+
+class DisplaySpy
```
Comment (MollieS) [inline spec/game_spec.rb]: If you want to keep these test doubles for reference, I would keep them in a branch and remove them from master as you're not using them.

File: spec/game_spec.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+    game.run
+
+    expect(board_spy).to have_received(:mark_board).at_least(2).times
+    expect(display_spy).to have_received(:display_board).with(grid).at_least(3).times
+    expect(display_spy).to have_received(:repeat?).at_least(1).time
+  end
+
+  it 'cannot pick the same coordinate twice' do
+    grid = []
```
Comment (MollieS) [inline spec/game_spec.rb]: Do you need the grid and the allow on the board to return it in every test? If so, could you use a before or let block to DRY this file up?

File: spec/game_spec.rb [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    allow(board_spy).to receive(:grid).and_return(grid)
+    allow(board_spy).to receive(:validate_choice).and_return(false)
+    allow(display_spy).to receive(:repeat?).and_return(false)
+
+    game = Game.new(display_spy, board_spy)
```
Comment (MollieS) [inline spec/game_spec.rb]: This duplication could also be removed by using a let or before block - in this instance, one option is better than the other... you could research the difference between the two.  It's subtle but also very important

File: spec/game_spec.rb (hunk 2 above)
Comment (backslashbaker) [reply spec/game_spec.rb]: How do I take into account the tests that require this to be true then false?

File: spec/game_spec.rb (hunk 7 above)
Comment (backslashbaker) [reply spec/game_spec.rb]: removed the grid variable as I wasn't actually using it

**Pull request merged**